public/app/features/explore/QueryRows.tsx

```diff
@@ -62,6 +68,20 @@ export const QueryRows = ({ exploreId }: Props) => {
     [onChange, queries]
   );
 
+  const hasQueryLibrary = config.featureToggles.queryLibrary || false;
+
+  useEffect(() => {
+    RowActionComponents.addExtraRenderAction((props) =>
+      hasQueryLibrary ? (
+        <QueryOperationAction
+          title={t('query-operation.header.save-to-query-library', 'Save to query library')}
+          icon="save"
+          onClick={() => onBeginQLSave(props.query?.refId)}
+        />
+      ) : null
+    );
+  }, [hasQueryLibrary]);
```

Comment:
Thank you for this, sorry I missed it. I checked out how Enterprise does it, and they just have an init function go off when you start the app that adds the action. See `initRecordedQueries` inside the Enterprise repository for more information.

Unfortunately, there's not an easy equivalent, because I still would like to encapsulate this inside Explore as much as possible. I moved the logic up to `ExplorePage`, the top level component for Explore, but it still would create redundant actions when you left Explore and came back. I decided to improve on the QueryActionComponent paradigm by adding another variable, a map, that you could set actions to. This way when Explore re-loads, it will re-add the action, but only by replacing the action with the same key. This is done alongside the existing array for actions, so it will not impact Enterprise and recorded queries. It only fires once whenever you enter Explore after navigating out of Explore.

As it is without this change, you can get the list of actions and iterate through it, but it's a list of JSX elements - I couldn't find an easy way to establish uniqueness on that list. 

It does mean that this PR is no longer isolated to just explore, but I think the keyed list of actions will have a lot of use if we build out extensions on query rows like we talked a bit about.  Having an init function fire once on app load can work well for Enterprise, but for more complicated cases inside core Grafana, we should let components that may load more than once be responsible for adding query actions without the risk of redundancy.

Reply:
One problem is that the save button will show up in all query rows (e.g. in Dashboards), but only after loading Explore page and will not do antyhing. Example: If I start with a dashboard - no button shows up. I go to Explore - button shows up and works. I go back to a dashbaord - the button shows up but does nothing.

https://github.com/user-attachments/assets/55a71fc9-49bc-4fd7-99a5-11ed7ea823d6

I think we should scope it down so the button is shown only in Explore. Once we expand it to other places we should use AppPlugin Extension and have the save modal being provided by the app (if installed).

One way of scoping it down would be to unregister the item on unmount, but definitely check the API with dashboards. It feels a bit hacky that there's a global singleton to inject additional options.

Reply:
>  It feels a bit hacky that there's a global singleton to inject additional options.

Unfortunately, theres not a lot of other options..

I really do like using QueryActionComponent for this and would like to continue this direction, because we are able to keep the add to QueryEditorRow to where it is relevant. I added the concept of scopes to the Map as well, so the QueryEditorRow will only add in the actions relevant to the app that is passed in. We're getting a little deep into adding stuff to this aspect of Grafana, but I think it's highly likely if other parts of core Grafana would want to add actions, they would run into the same problems needing scopes and uniqueness.